### ISSUE TITLE
Code coverage lib Tarpaulin doesn't work with non x86 architecture (ARM)

### ISSUE THREAD

**Issue by karolh2000** [pyrsia/pyrsia#901]
### Suggestion details

Since Tarpaulin is x86 specific and depends on ptrace it doesn’t work in a non x86 environment (ARM). From what I understand there is no plan to rewrite Tarpaulin not to depend on ptrace (at least nothin reliable). We might want to mention it in our documentation or consider a different framework. Open to discussion.

Steps:
0) ARM architecture OS
1) `cargo install cargo-tarpaulin`
2) `cargo tarpaulin --workspace`

Error:
```
  ...
  ...
  Finished test [unoptimized + debuginfo] target(s) in 1m 30s
  Executable unittests src/lib.rs (target/debug/deps/pyrsia-89e4e3bb75674fb0)
  Executable unittests src/lib.rs (target/debug/deps/pyrsia_blockchain_network-f515c23468739db6)
  Executable unittests src/main.rs (target/debug/deps/pyrsia-dab48345baae7e02)
  Executable unittests src/main.rs (target/debug/deps/pyrsia_node-f1e67295a8f66184)
Jul 27 13:45:14.982 ERROR cargo_tarpaulin::process_handling: Ptrace is not supported on this platform
Jul 27 13:45:14.983 ERROR cargo_tarpaulin: Failed to get test coverage! Error: Unsupported OS
Error: "Failed to get test coverage! Error: Unsupported OS"
```

**Comment by tiainen** [comment]
There's a WIP issue that switches to an llvm based tracing instead of relying on ptrace: https://github.com/xd009642/tarpaulin/issues/549. The work of this is tracked in a separate branch: https://github.com/xd009642/tarpaulin/tree/feat/llvm-profparser-integration. You can try it by installing from that git repo and branch:

    cargo install cargo-tarpaulin --git https://github.com/xd009642/tarpaulin --branch feat/llvm-profparser-integration

And then run tarpaulin as follows:

    cargo tarpaulin --engine Llvm -- --test-threads 1

**Comment by karolh2000** [comment]
@tiainen Correct, I did investigate this workaround (LLVM) yesterday but unfortunately wasn't able to get it to work on macOs, basically it hangs forever at the point shown in the logs below. However I used slightly different parameters so I'll give another shot.

```
test util::env_util::tests::test_value_present_trim ... ok
test util::keypair_util::tests::load_existing_keypair_with_wrong_size_fails ... ok
test util::keypair_util::tests::load_non_existing_keypair_generates_new_keypair_and_saves_it ... ok
test util::keypair_util::tests::load_non_existing_keypair_generates_new_keypair_but_does_not_save_it ... ok
test util::keypair_util::tests::saved_keypair_can_be_loaded ... ok
test verification_service::service::tests::test_verify_transaction ... ok

test result: ok. 99 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 6.25s

Jul 28 10:02:53.526  INFO cargo_tarpaulin::statemachine::instrumented: For binary: target/debug/deps/pyrsia-d54caeec18e83af3
Jul 28 10:02:53.526  INFO cargo_tarpaulin::statemachine::instrumented: Generated: pyrsia-d54caeec18e83af3_36196.profraw
```

**Comment by tiainen** [comment]
How long did you wait "forever"? I noticed too that the final process takes a while, but in the end it got through. Can't remember exactly how long I waited for it to complete, but it must have been less than 15 minutes.

**Comment by karolh2000** [comment]
It looks like it works with macOs (m1), still cannot make it work with docker/linux arm (macOs M1 client), not sure if it's even supported, I asked and we'll see what happens. https://github.com/xd009642/tarpaulin/issues/549#issuecomment-1198600914. It's nice to have so won't spend any more time on it.

**Comment by xlab** [comment]
> There's a WIP issue that switches to an llvm based tracing instead of relying on ptrace: [xd009642/tarpaulin#549](https://github.com/xd009642/tarpaulin/issues/549). The work of this is tracked in a separate branch: https://github.com/xd009642/tarpaulin/tree/feat/llvm-profparser-integration. You can try it by installing from that git repo and branch:
> 
> ```
> cargo install cargo-tarpaulin --git https://github.com/xd009642/tarpaulin --branch feat/llvm-profparser-integration
> ```
> 
> And then run tarpaulin as follows:
> 
> ```
> cargo tarpaulin --engine Llvm -- --test-threads 1
> ```

Thanks this worked for me! Finally.

I tried to use a x86_64 Docker but it would stuck during compilation lol.
Glad this can be done natively now.
